owners/src/parser.js

```diff
@@ -47,10 +62,140 @@ class OwnersParser {
     this.logger = logger || console;
   }
 
+  /**
+   * Parse a single owner declaration.
+   *
+   * TODO(rcebulko): Add support for teams.
+   *
+   * @private
+   * @param {!string} ownersPath OWNERS.yaml file path (for error reporting).
+   * @param {!string} owner owner username.
+   * @return {OwnersParserResult<string[]>} list of owners' usernames.
+   */
+  _parseOwnersLine(ownersPath, owner) {
+    const owners = [];
+    const errors = [];
+
+    if (owner[0] === '@') {
+      const lineResult = this._parseOwnersLine(ownersPath, owner.slice(1));
+
+      owners.push(...lineResult.result);
+      errors.push(
+        new OwnersParserError(
+          ownersPath,
+          `Ignoring unnecessary '@' in '${owner}'`
+        ),
+        ...lineResult.errors
+      );
+    } else if (owner.indexOf('/') !== -1) {
+      errors.push(
+        new OwnersParserError(
+          ownersPath,
+          `Failed to parse owner '${owner}'; team ownership not yet supported`
+        )
+      );
+    } else {
+      owners.push(owner);
+    }
+
+    return {result: owners, errors};
+  }
+
+  /**
+   * Parse a list of owners.
+   *
+   * @private
+   * @param {!string} ownersPath OWNERS.yaml file path (for error reporting).
+   * @param {string[]} ownersList list of owners.
+   * @return {OwnersParserResult<string[]>} list of owners' usernames.
+   */
+  _parseOwnersList(ownersPath, ownersList) {
+    const owners = [];
+    const errors = [];
+
+    ownersList.forEach(owner => {
+      const lineResult = this._parseOwnersLine(ownersPath, owner);
+      owners.push(...lineResult.result);
+      errors.push(...lineResult.errors);
+    });
+
+    return {result: owners, errors};
+  }
+
+  /**
+   * Parse an owners dictionary as a pattern-based rule.
+   *
+   * Note: All YAML parsed dictionaries have a single key-value pair; a dict not
+   * matching this will not be parsed correctly.
+   *
+   * @private
+   * @param {!string} ownersPath OWNERS.yaml file path (for error reporting).
+   * @param {!object} ownersDict dictionary with a pattern as the key and a list
+   * of owners as the value.
+   * @return {OwnersParserResult<PatternOwnersRule[]>} parsed OWNERS pattern
+   *     rule.
+   */
+  _parseOwnersDict(ownersPath, ownersDict) {
+    const [[pattern, ownersList]] = Object.entries(ownersDict);
+    const rules = [];
+    const errors = [];
+
+    // Treat each comma-separated subpattern as its own rule definition.
+    pattern.split(/\s*,\s*/).forEach(subpattern => {
+      const owners = [];
+      const isRecursive = subpattern.indexOf('**/') === 0;
+      if (isRecursive) {
+        subpattern = subpattern.slice(3);
```

Comment:
Supernit: you can also do `subpattern.slice('**/'.length)` to make this more clear, or even replace `'**/'` with a top-level const for that purpose

Reply:
I like this idea, done. Also ended up removing the need for the double-quotes you commented on above, since it became a format string